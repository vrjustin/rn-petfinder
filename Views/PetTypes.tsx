import React, {useEffect} from 'react';
import {
  Text,
  View,
  FlatList,
  StyleSheet,
  TouchableOpacity,
  ImageBackground,
  Dimensions,
} from 'react-native';
import {useNavigation} from '@react-navigation/native';
import {useSelector, useDispatch} from 'react-redux';
import {setPetTypes, selectPetTypes} from '../reducers/petTypesReducer';
import {selectFavorites} from '../reducers/animalsReducer';
import apiService from '../services/apiService';
import {PetType, petTypeImages} from '../models/PetType';
import FontAwesomeIcon from 'react-native-vector-icons/FontAwesome';
import GlobalStyles from './Styles/GlobalStyles';
import {Routes} from '../navigation/Routes';

const screenWidth = Dimensions.get('window').width;

const PetTypes: React.FC = () => {
  const navigation = useNavigation();
  const dispatch = useDispatch();
  const petTypes = useSelector(selectPetTypes);
  const globalStyles = GlobalStyles();
  const favoriteAnimals = useSelector(selectFavorites);
  const haveFavorites = favoriteAnimals.length > 0;

  useEffect(() => {
    const fetchPetTypesData = async () => {
      try {
        const typesData = await apiService.getPetTypes();
        dispatch(setPetTypes(typesData?.types ?? []));
      } catch (error) {
        console.error('Failed to fetch pet types:', error);
      }
    };
    fetchPetTypesData();
  }, [dispatch]);

  const handlePetTypeSelection = (petType: PetType) => {
    navigation.navigate(Routes.Breeds, {petType: petType});
  };

  const handleFavoriteSelectionNavigation = () => {
    const favType: PetType = {
      name: 'Favorite',
      coats: [],
      colors: [],
      genders: [],
      _links: {
        self: {href: ''},
        breeds: {href: ''},
      },
    };
    navigation.navigate(Routes.Animals, {petType: favType});
  };

  const renderPetType = ({item}: {item: PetType}) => (
    <TouchableOpacity onPress={() => handlePetTypeSelection(item)}>
      <View style={styles.gridItemContainer}>
        <ImageBackground
          source={
            petTypeImages[item.name]
              ? petTypeImages[item.name]
              : require('../resources/black-1869685_1280.jpg')
          }
          style={styles.itemBackground}
          imageStyle={styles.gridItemImage}>
          <View style={styles.item}>
            <FontAwesomeIcon
              name="paw"
              size={20}
              color="#fff"
              style={styles.icon}
            />
            <Text style={styles.text}>{item.name}</Text>
          </View>
        </ImageBackground>
      </View>
    </TouchableOpacity>
  );

  const renderFavoritesTouchable = () => {
    return haveFavorites ? (
      <TouchableOpacity onPress={handleFavoriteSelectionNavigation}>
        <View
          style={{
            backgroundColor: 'pink',
            height: 38,
            flexDirection: 'row',
            justifyContent: 'center',
            alignItems: 'center',
            marginHorizontal: 20,
            marginBottom: 8,
            borderRadius: 8,
          }}>
          <FontAwesomeIcon
            name={'heart'}
            size={20}
            color={'white'}
            style={styles.icon}
          />
          <Text style={{marginRight: 8}}>Favorites</Text>
          <FontAwesomeIcon
            name={'heart'}
            size={20}
            color={'white'}
            style={styles.icon}
          />
        </View>
      </TouchableOpacity>
    ) : null;
  };

  if (petTypes.length === 0) {
    return <Text>Loading...</Text>;
  }

  return (
    <View style={globalStyles.container}>
      <Text style={globalStyles.header}>
        Select an Animal Type to Get Started:
      </Text>
      {renderFavoritesTouchable()}
      <View style={styles.gridContainer}>
        <FlatList
          data={petTypes}
          renderItem={renderPetType}
          keyExtractor={item => item.name}
          numColumns={2}
        />
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  itemBackground: {
    flex: 1,
    margin: 10,
    height: 150,
    justifyContent: 'center',
    borderRadius: 10,
    overflow: 'hidden',
  },
  item: {
    flexDirection: 'row',
    alignItems: 'flex-end',
    justifyContent: 'center',
    padding: 16,
    flex: 1,
  },
  gridContainer: {
    flex: 1,
    alignItems: 'center',
  },
  gridItemContainer: {
    flex: 1,
    aspectRatio: 1,
    margin: 4,
    width: (screenWidth - 32) / 2,
  },
  text: {
    fontSize: 16,
    color: '#fff',
    marginLeft: 10,
  },
  icon: {
    marginRight: 10,
  },
  gridItemImage: {
    borderRadius: 8,
  },
});

export default PetTypes;
