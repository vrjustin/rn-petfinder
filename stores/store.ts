import {combineReducers, configureStore} from '@reduxjs/toolkit';
import petTypesReducer from '../reducers/petTypesReducer';
import petBreedsReducer from '../reducers/petBreedsReducer';
import animalsReducer from '../reducers/animalsReducer';
import searchParamsReducer from '../reducers/searchParamsReducer';

const createDebugger = require('redux-flipper').default;

const rootReducer = combineReducers({
  petTypes: petTypesReducer,
  petBreeds: petBreedsReducer,
  animals: animalsReducer,
  searchParameters: searchParamsReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

const store = configureStore({
  reducer: rootReducer,
  middleware: getDefaultMiddleware =>
    __DEV__
      ? getDefaultMiddleware({serializableCheck: false}).concat(
          createDebugger(),
        )
      : getDefaultMiddleware({serializableCheck: false}),
});

export default store;
