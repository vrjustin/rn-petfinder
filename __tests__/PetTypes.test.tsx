import React from 'react';
import {Provider} from 'react-redux';
import store from '../stores/store';
import renderer from 'react-test-renderer';
import PetTypes from '../Views/PetTypes';

jest.mock('@react-native-async-storage/async-storage', () => ({
  setItem: jest.fn(),
  getItem: jest.fn(),
}));

jest.mock('redux-persist', () => ({
  ...jest.requireActual('redux-persist'),
  persistStore: jest.fn(),
}));

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
    setOptions: jest.fn(),
  }),
}));

jest.mock('../services/apiService', () => ({
  getPetTypes: jest.fn(() => ({
    types: [
      {
        name: 'Dog',
        coats: ['Short'],
        colors: ['Brown'],
        genders: ['Male'],
        _links: {self: {href: ''}, breeds: {href: ''}},
      },
      {
        name: 'Cat',
        coats: ['Long'],
        colors: ['Black'],
        genders: ['Female'],
        _links: {self: {href: ''}, breeds: {href: ''}},
      },
      {
        name: 'Bird',
        coats: [],
        colors: ['Green'],
        genders: ['Unknown'],
        _links: {self: {href: ''}, breeds: {href: ''}},
      },
    ],
  })),
}));

let renderedPetTypesTree: any;

describe('PetTypes', () => {
  it('renders correctly when no petTypes local and is loading', () => {
    renderedPetTypesTree = renderer.create(
      <Provider store={store}>
        <PetTypes />
      </Provider>,
    );
    expect(renderedPetTypesTree.toJSON()).toMatchSnapshot();
  });
});
