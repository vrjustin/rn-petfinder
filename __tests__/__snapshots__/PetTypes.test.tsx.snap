// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PetTypes renders correctly when no petTypes local and is loading 1`] = `
<Text>
  Loading...
</Text>
`;

exports[`PetTypes renders correctly when petTypes are fetched 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "white",
        "flex": 1,
        "padding": 0,
      }
    }
  >
    <Text
      style={
        {
          "color": "black",
          "fontSize": 20,
          "fontWeight": "bold",
          "margin": 20,
        }
      }
    >
      Select an Animal Type to Get Started:
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="PetTypes-OrganizationsButton"
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "grey",
            "borderRadius": 8,
            "flexDirection": "row",
            "height": 52,
            "justifyContent": "center",
            "marginBottom": 8,
            "marginHorizontal": 20,
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "white",
                "fontSize": 20,
              },
              {
                "fontSize": 24,
                "marginRight": 10,
              },
              {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          
        </Text>
        <Text
          style={
            {
              "color": "white",
              "marginRight": 8,
            }
          }
        >
          Adoption Shelters
        </Text>
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "white",
                "fontSize": 20,
              },
              {
                "fontSize": 24,
                "marginRight": 10,
              },
              {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <RCTScrollView
        data={
          [
            {
              "_links": {
                "breeds": {
                  "href": "",
                },
                "self": {
                  "href": "",
                },
              },
              "coats": [
                "Short",
              ],
              "colors": [
                "Brown",
              ],
              "displayName": "Dog",
              "genders": [
                "Male",
              ],
              "name": "Dog",
            },
            {
              "_links": {
                "breeds": {
                  "href": "",
                },
                "self": {
                  "href": "",
                },
              },
              "coats": [
                "Long",
              ],
              "colors": [
                "Black",
              ],
              "displayName": "Cat",
              "genders": [
                "Female",
              ],
              "name": "Cat",
            },
            {
              "_links": {
                "breeds": {
                  "href": "",
                },
                "self": {
                  "href": "",
                },
              },
              "coats": [],
              "colors": [
                "Green",
              ],
              "displayName": "Bird",
              "genders": [
                "Unknown",
              ],
              "name": "Bird",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                {
                  "flexDirection": "row",
                }
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
                testID="PetTypes-PetType-Dog-button"
              >
                <View
                  style={
                    {
                      "aspectRatio": 1,
                      "flex": 1,
                      "margin": 4,
                      "width": 359,
                    }
                  }
                >
                  <View
                    accessibilityIgnoresInvertColors={true}
                    style={
                      {
                        "borderRadius": 10,
                        "flex": 1,
                        "justifyContent": "center",
                        "margin": 10,
                      }
                    }
                  >
                    <Image
                      source={
                        {
                          "default": "",
                        }
                      }
                      style={
                        [
                          {
                            "bottom": 0,
                            "left": 0,
                            "position": "absolute",
                            "right": 0,
                            "top": 0,
                          },
                          {
                            "height": undefined,
                            "width": undefined,
                          },
                          {
                            "borderRadius": 8,
                          },
                        ]
                      }
                    />
                    <View
                      style={
                        {
                          "alignItems": "flex-end",
                          "flex": 1,
                          "flexDirection": "row",
                          "justifyContent": "center",
                          "padding": 16,
                        }
                      }
                    >
                      <Text
                        allowFontScaling={false}
                        selectable={false}
                        style={
                          [
                            {
                              "color": "#fff",
                              "fontSize": 20,
                            },
                            {
                              "fontSize": 24,
                              "marginRight": 10,
                            },
                            {
                              "fontFamily": "FontAwesome",
                              "fontStyle": "normal",
                              "fontWeight": "normal",
                            },
                            {},
                          ]
                        }
                      >
                        
                      </Text>
                      <Text
                        style={
                          {
                            "color": "#fff",
                            "fontSize": 16,
                            "marginLeft": 10,
                          }
                        }
                      >
                        Dog
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
                testID="PetTypes-PetType-Cat-button"
              >
                <View
                  style={
                    {
                      "aspectRatio": 1,
                      "flex": 1,
                      "margin": 4,
                      "width": 359,
                    }
                  }
                >
                  <View
                    accessibilityIgnoresInvertColors={true}
                    style={
                      {
                        "borderRadius": 10,
                        "flex": 1,
                        "justifyContent": "center",
                        "margin": 10,
                      }
                    }
                  >
                    <Image
                      source={
                        {
                          "default": "",
                        }
                      }
                      style={
                        [
                          {
                            "bottom": 0,
                            "left": 0,
                            "position": "absolute",
                            "right": 0,
                            "top": 0,
                          },
                          {
                            "height": undefined,
                            "width": undefined,
                          },
                          {
                            "borderRadius": 8,
                          },
                        ]
                      }
                    />
                    <View
                      style={
                        {
                          "alignItems": "flex-end",
                          "flex": 1,
                          "flexDirection": "row",
                          "justifyContent": "center",
                          "padding": 16,
                        }
                      }
                    >
                      <Text
                        allowFontScaling={false}
                        selectable={false}
                        style={
                          [
                            {
                              "color": "#fff",
                              "fontSize": 20,
                            },
                            {
                              "fontSize": 24,
                              "marginRight": 10,
                            },
                            {
                              "fontFamily": "FontAwesome",
                              "fontStyle": "normal",
                              "fontWeight": "normal",
                            },
                            {},
                          ]
                        }
                      >
                        
                      </Text>
                      <Text
                        style={
                          {
                            "color": "#fff",
                            "fontSize": 16,
                            "marginLeft": 10,
                          }
                        }
                      >
                        Cat
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                {
                  "flexDirection": "row",
                }
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
                testID="PetTypes-PetType-Bird-button"
              >
                <View
                  style={
                    {
                      "aspectRatio": 1,
                      "flex": 1,
                      "margin": 4,
                      "width": 359,
                    }
                  }
                >
                  <View
                    accessibilityIgnoresInvertColors={true}
                    style={
                      {
                        "borderRadius": 10,
                        "flex": 1,
                        "justifyContent": "center",
                        "margin": 10,
                      }
                    }
                  >
                    <Image
                      source={
                        {
                          "default": "",
                        }
                      }
                      style={
                        [
                          {
                            "bottom": 0,
                            "left": 0,
                            "position": "absolute",
                            "right": 0,
                            "top": 0,
                          },
                          {
                            "height": undefined,
                            "width": undefined,
                          },
                          {
                            "borderRadius": 8,
                          },
                        ]
                      }
                    />
                    <View
                      style={
                        {
                          "alignItems": "flex-end",
                          "flex": 1,
                          "flexDirection": "row",
                          "justifyContent": "center",
                          "padding": 16,
                        }
                      }
                    >
                      <Text
                        allowFontScaling={false}
                        selectable={false}
                        style={
                          [
                            {
                              "color": "#fff",
                              "fontSize": 20,
                            },
                            {
                              "fontSize": 24,
                              "marginRight": 10,
                            },
                            {
                              "fontFamily": "FontAwesome",
                              "fontStyle": "normal",
                              "fontWeight": "normal",
                            },
                            {},
                          ]
                        }
                      >
                        
                      </Text>
                      <Text
                        style={
                          {
                            "color": "#fff",
                            "fontSize": 16,
                            "marginLeft": 10,
                          }
                        }
                      >
                        Bird
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</RCTSafeAreaView>
`;
